lgc/patch/Continufy.cpp

```diff
@@ -0,0 +1,202 @@
+ *
+ *  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ *  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ *  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ *  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ *  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ *  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ *  SOFTWARE.
+ *
+ **********************************************************************************************************************/
+/**
+ ***********************************************************************************************************************
+ * @file  Continufy.cpp
+ * @brief LLPC source file: contains implementation of class lgc::Continufy.
+ * This pass translates indirect call into cps.await call, which will be lowered into continuation call.
+ ***********************************************************************************************************************
+ */
+#include "lgc/patch/Continufy.h"
+#include "lgccps/LgcCpsDialect.h"
+#include "lgcrt/LgcRtDialect.h"
+#include "lgc/Builder.h"
+#include "lgc/LgcDialect.h"
+#include "lgc/patch/Patch.h"
+#include "lgc/state/PalMetadata.h"
+#include "llvm/Support/Debug.h"
+
+#define DEBUG_TYPE "lgc-continufy"
+
+using namespace llvm;
+using namespace lgc;
+using namespace lgc::cps;
+
+namespace lgc {
+using RtStage = rt::RayTracingShaderStage;
+
+static Function *insertCpsArguments(Function &fn) {
+  // Mutate function arguments, add ({} %state, %rcr).
+  LLVMContext &context = fn.getContext();
+  SmallVector<Type *> argTys = {StructType::get(context, {}), IntegerType::get(context, 32)};
+  auto *fnTy = fn.getFunctionType();
+  argTys.append(fnTy->params().begin(), fnTy->params().end());
+
+  auto *newFn = mutateFunctionArguments(fn, Type::getVoidTy(context), argTys, fn.getAttributes());
+
+  fn.replaceAllUsesWith(newFn);
+  for (unsigned idx = 0; idx < fn.arg_size(); idx++) {
+    Value *oldArg = fn.getArg(idx);
+    Value *newArg = newFn->getArg(idx + 2);
+    newArg->setName(oldArg->getName());
+    oldArg->replaceAllUsesWith(newArg);
+  }
+  newFn->getArg(0)->setName("state");
+  newFn->getArg(1)->setName("rcr");
+  return newFn;
+}
+
+/// Return the CPS levels mask of the ray-tracing stages that the input stage will return to.
+static unsigned getReturnedLevels(int stage) {
+  // Traversal will return to RGS or CHS/MISS.
+  if (stage == -1)
+    return 1u << (unsigned)CpsLevel::RayGen | 1u << (unsigned)CpsLevel::ClosestHit_Miss_Callable;
+
+  RtStage rtStage = static_cast<RtStage>(stage);
+  switch (rtStage) {
+  case RtStage::ShaderStageRayGeneration:
+    llvm_unreachable("Raygen shader should not arrive here.");
+  case RtStage::ShaderStageClosestHit:
+  case RtStage::ShaderStageMiss:
+    // Traversal
+    return (1u << (unsigned)CpsLevel::Traversal);
```

Comment:
Aren't we directly jumping back to RGS/CHS/Miss from CHS and Miss?

When calling CHS/Miss, Traversal obtains the RGS/CHS/Miss return address from system data, and passes it to CHS/Miss.

Reply:
In the real continuations workflow, yes. With the continufy pass, no. It will keep the “real” function stack of
RayGen → TraceRay → CHS/Miss
and return the same way.

Reply:
Oh, I see. That warrants a comment though :)

Reply:
I am not sure what kind of comment would be more helpful here. I just added some comment to help people be more aware of the usage of Continufy pass.